src/main/sampling/Sample.cpp

```diff
@@ -182,6 +182,122 @@ namespace lsp
+                                        dptr[c * chunk_size + i] += sptr[ (size_t)(c*stretch_ratio*chunk_size + i) ] * gain;
+                                    }
+                                }
+                            }
+                        } else if (stretch_samples < 0) {
+                            /**
+                             *  STRETCH UP
+                             */
+                            float  stretch_ratio = (float)n_chunks_dst/n_chunks;
+
+                            for (size_t c = 0; c < n_chunks_dst; c++) {
+                                for (size_t i = 0; i < chunk_size; i++) {
+                                    if (c == 0) {
+                                        float gain = (i > chunk_size / 2) && n_chunks_dst > 1 ? sqrtf(((float)chunk_size-i)/chunk_size) : 1;
+                                        dptr[c * chunk_size + i] += sptr[ (size_t)(c / stretch_ratio * chunk_size + i)] * gain;
+                                        float ogain = (i > chunk_size / 2) ? sqrtf(((float)chunk_size-i)/chunk_size) : sqrtf((float)i/chunk_size);
+                                        dptr[ (size_t)(c * chunk_size + chunk_size/2 + i) ] += sptr[ (size_t)((c+0.5f) / stretch_ratio * chunk_size + i) ] * ogain;
+                                    } else if(c < n_chunks_dst-1) {
+                                        float gain = (i > chunk_size / 2) ? sqrtf(((float)chunk_size-i)/(chunk_size/2)) : sqrtf((float)i/(chunk_size/2));
+                                        dptr[c * chunk_size + i] += sptr[ (size_t)(c / stretch_ratio * chunk_size + i) ]*gain;
+                                        dptr[ (size_t)(c * chunk_size + chunk_size/2 + i) ] += sptr[ (size_t)((c+0.5f) / stretch_ratio * chunk_size + i) ] * gain;
+                                    } else {
+                                        float gain = (i < chunk_size / 2) ? sqrtf((float)i/chunk_size) : 1;
+                                        dptr[c * chunk_size + i] += sptr[ (size_t)(c / stretch_ratio * chunk_size + i) ] * gain;
+                                    }
+                                }
+                            }
+                        } else {
+                            for (size_t i = 0; i < length; i++) {
+                                dptr[i] = sptr[i];
+                            }
+                        }
+                    }
+                    else
+                        dsp::fill_zero(dptr, max_length);
+
+                    dptr           += max_length;
+                }
+
+                // Destroy previously allocated data
+                if (vBuffer != NULL)
+                    free(vBuffer);
+            }
+            else
+                dsp::fill_zero(buf, max_length * channels);
+
+            vBuffer         = buf;
+            nLength         = length;
+            nMaxLength      = max_length;
+            nChannels       = channels;
+            return true;
+        }
```

Comment:
It would be also nice to have the unit test `src/test/utest/sampling/sample.cpp` updated that:
- checks the proper function of the method under different conditions;
- allows to verify proper memory utilization by launching the test under `valgrind`;
- allows to verify that no memory corruption is possible also by launching the test under `valgrind`.

Reply:
I will look into that a bit later. If you wanna get a better understanding of how this works you could try running [lsp-plugins-sampler/PR-4](https://github.com/lsp-plugins/lsp-plugins-sampler/pull/4) with this branch as a `lsp-dsp-units` modules.

Reply:
I understand how the algorithm works. But we need to ensure that some certain conditions won't break the behaviour or make the algorithm overrun some memory buffers.

Reply:
Could you tell me how do you output stuff to the terminal. I'm using `std::cout` right now. :D

Reply:
Use debug/trace functions from <lsp-plug.in/common/debug.h>. They work like ordinary `printf`.

Reply:
You also can easily debug your code if you use load the project in Eclipse+CDT.

Reply:
I've implemented the features you've suggested. Would really appreciate some reviews on the update so we could make sure this feature works well.

I'm gonna work on these tests now. What about debugging in a terminal? I'm more of a vim user TBH :D

P.S.
Is there any developer chat for the lsp-plugins project?

Reply:
Okay, I'll look again within few days at it.
You can join the telegram chat related to development of LSP Plugins. Here's the invite link: https://t.me/+IMdUkoBBD801tppt